go-controller/pkg/libovsdbops/db_index.go

```diff
@@ -0,0 +1,259 @@
+//
+//	const (
+//		ipFamilyKey = "ipFamily"
+//		Controller = "controller-1"
+//		ObjectType = "AddressSetType"
+//	)
+//
+//	type addrSetIDs struct {
+//		NamespaceName string
+//      ipFamily string
+//	}
+//
+//	func (a *addrSetIDs) GetExternalIDs() map[string]string {
+//		return map[string]string{
+//			SecondaryIndexKey: a.NamespaceName,
+//			ipFamilyKey:      a.ipFamily,
+//		}
+//	}
+//	func (a *addrSetIDs) GetUniqueID() string {
+//		uniqueID := a.NamespaceName
+//		if a.ipFamily != "" {
+//			uniqueID += ":" + a.ipFamily
+//		}
+//		return uniqueID
+//	}
+//
+// This struct may be used to create DbIndex
+//	dbIndex := NewDbIndex[*addrSetIDs](Controller, ObjectType, &addrSetIDs{
+//		NamespaceName: "namespace1_name1",
+//		ipFamily: "ipv4",
+//	})
+//
+// This index may be used to create db object
+//	as := &nbdb.AddressSet{
+//		Name:        dbIndex.GetUniqueID(),
+//		ExternalIDs: dbIndex.GetExternalIDs(),
+//	}
+//
+// The following ExternalIDs will be created for this example
+// {
+//		"k8s.ovn.org/controller": "controller-1"
+//		"k8s.ovn.org/object-type": "AddressSetType"
+//		"k8s.ovn.org/name": "namespace1_name1"
+//		"k8s.ovn.org/id": "AddressSetType.controller-1.k8s.ovn.org/namespace1_name1:ipv4"
+//		"AddressSetType.controller-1.k8s.ovn.org/ipFamily": "ipv4"
+// }
+// AddressSet.Name = AddressSetType.controller-1.k8s.ovn.org/namespace1_name1:ipv4
+//
+// Check addressset.DbIDs object for more examples.
+// To ensure correct usage of DbIndex, and field verification, always use NewDbIndex to create the object.
+type DbIndex[T DbObjectIDs] struct {
+	// OwnerController specifies which controller owns the object.
+	// Controller should only change objects it owns, make sure to always set this field.
+	OwnerController string
+	// OwnerObjectType should identify object type that owns db object, e.g. Pod, NetworkPolicy, etc.
+	// Controller handlers should use unique OwnerTypes to only change objects they manage.
+	// Make sure to always set this field.
+	OwnerObjectType string
+	// ownerKey is composed internally and is constructed based on OwnerController and OwnerObjectType.
+	// It creates a context, in which internal DbObjectIDs must be unique.
+	ownerKey string
+	// ObjectIDs are ids that may be different for every combination of OwnerController and OwnerObjectType.
+	// These ids should uniquely identify db object with the same OwnerController and OwnerObjectType.
+	// It may even be empty, in case there is only 1 object of given type in the whole cluster.
+	ObjectIDs T
+}
+
+// DbObjectIDs is used to define internal object ids in addition to DbIndex.OwnerObjectType and DbIndex.OwnerController
+// (let's call this combination an Owner). These ids will be used to set db object ExternalIDs.
+// The only requirement for DbObjectIDs is to make sure they uniquely identify every object with given Owner.
+type DbObjectIDs interface {
```

Comment:
I was hoping that IDs of object types would be registered in kind of a central table instead of being provided externally (through this interface). I think it provides more flexibility than we need (and probably want).